--- mpg123-1.22.4/configure.ac
+++ mpg123-1.22.4-edit/configure.ac
@@ -21,7 +21,10 @@
 AC_CONFIG_AUX_DIR(build)
 AC_CONFIG_SRCDIR(doc)
 AC_CONFIG_MACRO_DIR([m4])
+AM_MAINTAINER_MODE
+cflags_save="$CFLAGS"
 AC_CANONICAL_HOST
+CFLAGS="$cflags_save"
 
 dnl Version 1.7 of automake is recommended
 AM_INIT_AUTOMAKE(1.7)
@@ -75,9 +78,11 @@
 	AC_MSG_WARN( Use -static in LDFLAGS for all-static linking! Your compiler may blow up on that -all-static. )
 fi
 
+cflags_save="$CFLAGS"
 AM_PROG_AS
 AC_PROG_CC
 AC_PROG_CPP
+CFLAGS="$cflags_save"
 AC_PROG_INSTALL
 dnl "Checking for egrep is broken after removal of libltdl stuff... checks use $EGREP, so searching it here."
 AC_PROG_EGREP
@@ -161,8 +166,9 @@
 
 dnl Configure libtool
 
+#LT_INIT([win32-dll])
 AC_LIBTOOL_WIN32_DLL
-AM_PROG_LIBTOOL
+AC_PROG_LIBTOOL
 
 if test x"$modules" = xdisabled
 then
@@ -1141,7 +1147,7 @@
 
 # Checks for maths libraries.
 AC_CHECK_LIB([m], [sqrt])
-AC_CHECK_LIB([mx], [powf])
+#AC_CHECK_LIB([mx], [powf])
 
 # attempt to make the signal stuff work... also with GENERIC - later
 #if test x"$ac_cv_header_sys_signal_h" = xyes; then
@@ -1411,6 +1417,12 @@
     esac
   ;;
 esac
+case $host in
+  x86_64-apple-darwin*)
+    ADD_CFLAGS="$ADD_CFLAGS -mmacosx-version-min=10.5"
+    ADD_LDFLAGS="$ADD_LDFLAGS -mmacosx-version-min=10.5"
+  ;;
+esac
 
 # Use yasm instead of the default assembler for AVX sources
 if test "x$use_yasm_for_avx" = "xyes"; then
@@ -2145,19 +2157,19 @@
 case $with_optimization in
   0)
     # No Optimizations
-    CFLAGS="$CFLAGS"
+    CFLAGS="-g -fno-common $CFLAGS"
   ;;
   1)
-    CFLAGS="-O $CFLAGS"
+    CFLAGS="-O -fno-common $CFLAGS"
   ;;
   2)
-    CFLAGS="-O2 -fomit-frame-pointer -funroll-all-loops -finline-functions -ffast-math $CFLAGS"
+    CFLAGS="-O2 -fomit-frame-pointer -ffast-math -fno-common $CFLAGS"
   ;;
   3)
-    CFLAGS="-O3 -fomit-frame-pointer -funroll-all-loops -finline-functions -ffast-math $CFLAGS"
+    CFLAGS="-O3 -fomit-frame-pointer -funroll-all-loops -finline-functions -ffast-math -fno-common $CFLAGS"
   ;;
   4)
-    CFLAGS="-O4 -fomit-frame-pointer -funroll-all-loops -finline-functions -ffast-math $CFLAGS"
+    CFLAGS="-O4 -fomit-frame-pointer -funroll-all-loops -finline-functions -ffast-math -fno-common $CFLAGS"
   ;;
   *)
   	AC_MSG_ERROR([Unknown optimizations level '$with_optimization'])
