--- mpg123-1.22.4/src/libmpg123/mpg123lib_intern.h
+++ mpg123-1.22.4-edit/src/libmpg123/mpg123lib_intern.h
@@ -287,12 +287,21 @@
 #include "frame.h"
 
 /* fr is a mpg123_handle* by convention here... */
+#if 0
 #define NOQUIET  (!(fr->p.flags & MPG123_QUIET))
 #define VERBOSE  (NOQUIET && fr->p.verbose)
 #define VERBOSE2 (NOQUIET && fr->p.verbose > 1)
 #define VERBOSE3 (NOQUIET && fr->p.verbose > 2)
 #define VERBOSE4 (NOQUIET && fr->p.verbose > 3)
 #define PVERB(mp, level) (!((mp)->flags & MPG123_QUIET) && (mp)->verbose >= (level))
+#else
+#define NOQUIET 0
+#define VERBOSE 0
+#define VERBOSE2 0
+#define VERBOSE3 0
+#define VERBOSE4 0
+#define PVERB(mp, level) 0
+#endif
 
 int decode_update(mpg123_handle *mh);
 /* residing in format.c  */
