From b1524720b97586f551099de847d7f33b9d165b84 Mon Sep 17 00:00:00 2001
From: sezero <sezero@users.sourceforge.net>
Date: Sat, 10 Nov 2018 16:51:01 +0300
Subject: [PATCH] fix typo introduced in commit 8d73daa

---
 lib/sharedbook.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/sharedbook.c b/lib/sharedbook.c
index 8d73daa8..eeae9305 100644
--- a/lib/sharedbook.c
+++ b/lib/sharedbook.c
@@ -68,7 +68,7 @@ float _float32_unpack(long val){
     exp=63;
   }
   if (exp<-63){
-    exp-63;
+    exp=-63;
   }
   return(ldexp(mant,exp));
 }
