djgpp -- remember to add -ffast-math to CFLAGS

diff -u vorbis/lib/os.h~ vorbis/lib/os.h
--- vorbis/lib/os.h~
+++ vorbis/lib/os.h
@@ -38,10 +38,6 @@
 #    define STIN static
 #  endif
 
-#ifdef DJGPP
-#  define rint(x)   (floor((x)+0.5f))
-#endif
-
 #ifndef M_PI
 #  define M_PI (3.1415926536f)
 #endif
@@ -59,7 +55,7 @@ extern double rint(double);
 #pragma aux rint parm [8087] value [8087] = "frndint";
 #endif
 
-#ifdef __EMX__
+#if defined(__EMX__) || defined(__DJGPP__)
 # define rint vorbis_rint
 static __inline double vorbis_rint (double x) {
   double retval;
diff -u /dev/null vorbis/lib/Makefile.djgpp
--- /dev/null
+++ vorbis/lib/Makefile.djgpp
@@ -0,0 +1,34 @@
+CFLAGS  = -Wall -O2 -ffast-math -fsigned-char -I../include
+#CFLAGS  += -DHAVE_INTTYPES_H -DHAVE_STDINT_H
+LIBNAME = libvorbis.a
+LIBNAME2= libvorbisfile.a
+
+ifeq ($(CROSS),1)
+PREFIX=i586-pc-msdosdjgpp-
+else
+PREFIX=
+endif
+CC      = $(PREFIX)gcc
+AR      = $(PREFIX)ar
+RANLIB  = $(PREFIX)ranlib
+
+OBJ = mdct.o smallft.o block.o envelope.o window.o lsp.o lpc.o analysis.o synthesis.o psy.o info.o floor1.o floor0.o res0.o mapping0.o registry.o codebook.o sharedbook.o bitrate.o
+# lookup.o
+OBJ2= vorbisfile.o
+
+all: $(LIBNAME) $(LIBNAME2)
+
+%.o: %.c
+	$(CC) $(CFLAGS) -c -o $@ $<
+
+$(LIBNAME): $(OBJ)
+	$(AR) cru $(LIBNAME) $(OBJ)
+	$(RANLIB) $(LIBNAME)
+$(LIBNAME2): $(OBJ2)
+	$(AR) cru $(LIBNAME2) $(OBJ2)
+	$(RANLIB) $(LIBNAME2)
+
+clean:
+	rm -f *.o
+distclean: clean
+	rm -f *.a

