diff -u flac-1.3.x-dec/src/libFLAC/stream_decoder.c~ flac-1.3.x-dec/src/libFLAC/stream_decoder.c
--- flac-1.3.x-dec/src/libFLAC/stream_decoder.c~
+++ flac-1.3.x-dec/src/libFLAC/stream_decoder.c
@@ -1252,7 +1252,7 @@ FILE *get_binary_stdin_(void)
 	 */
 #if defined _MSC_VER || defined __MINGW32__
 	_setmode(_fileno(stdin), _O_BINARY);
-#elif defined __EMX__
+#elif defined __EMX__ || defined __OS2__ || defined __NT__ || defined __DJGPP__
 	setmode(fileno(stdin), O_BINARY);
 #endif
 
diff -u flac-1.3.x-dec/src/libFLAC/stream_encoder.c~ flac-1.3.x-dec/src/libFLAC/stream_encoder.c
--- flac-1.3.x-dec/src/libFLAC/stream_encoder.c~
+++ flac-1.3.x-dec/src/libFLAC/stream_encoder.c
@@ -4304,7 +4304,7 @@ FILE *get_binary_stdout_(void)
 	 */
 #if defined _MSC_VER || defined __MINGW32__
 	_setmode(_fileno(stdout), _O_BINARY);
-#elif defined __EMX__
+#elif defined __EMX__ || defined __OS2__ || defined __NT__ || defined __DJGPP__
 	setmode(fileno(stdout), O_BINARY);
 #endif
 
