diff -r 23000b73ce4e test/testautomation_timer.c
--- a/test/testautomation_timer.c	Mon Aug 14 23:45:06 2017 -0700
+++ b/test/testautomation_timer.c	Tue Aug 15 18:02:10 2017 +0300
@@ -108,12 +108,19 @@ Uint32 SDLCALL _timerTestCallback(Uint32
 {
    _timerCallbackCalled = 1;
 
+// [Digi, 2016]:
+// SDL_AddTimer() -> SDL_TimerInit() -> SDL_CreateThread() - timer thread that
+// have called this function was created with _beginthread() of SDL2.DLL's RTL, not
+// not our RTL! So we can't use any LIBC functions here! Hm... what about other
+// OSes/compilers ?!
+#if !defined(__WATCOMC__)
    if (_paramCheck != 0) {
        SDLTest_AssertCheck(param != NULL, "Check param pointer, expected: non-NULL, got: %s", (param != NULL) ? "non-NULL" : "NULL");
        if (param != NULL) {
           SDLTest_AssertCheck(*(int *)param == _paramValue, "Check param value, expected: %i, got: %i", _paramValue, *(int *)param);
        }
    }
+#endif
 
    return 0;
 }

