diff -up libwebp-1.0.1/src/dsp/dec.c~ libwebp-1.0.1/src/dsp/dec.c
--- libwebp-1.0.1/src/dsp/dec.c~
+++ libwebp-1.0.1/src/dsp/dec.c
@@ -255,13 +255,12 @@ VP8PredFunc VP8PredLuma16[NUM_B_DC_MODES
 #if !WEBP_NEON_OMIT_C_CODE
 static void VE4_C(uint8_t* dst) {    // vertical
   const uint8_t* top = dst - BPS;
-  const uint8_t vals[4] = {
-    AVG3(top[-1], top[0], top[1]),
-    AVG3(top[ 0], top[1], top[2]),
-    AVG3(top[ 1], top[2], top[3]),
-    AVG3(top[ 2], top[3], top[4])
-  };
+  uint8_t vals[4];
   int i;
+  vals[0] = AVG3(top[-1], top[0], top[1]);
+  vals[1] = AVG3(top[ 0], top[1], top[2]);
+  vals[2] = AVG3(top[ 1], top[2], top[3]);
+  vals[3] = AVG3(top[ 2], top[3], top[4]);
   for (i = 0; i < 4; ++i) {
     memcpy(dst + i * BPS, vals, sizeof(vals));
   }
